src/include/ipc/topology.h

```diff
@@ -65,8 +65,8 @@ struct sof_ipc_comp {
 	uint32_t pipeline_id;
 	uint32_t core;
 
-	/* reserved for future use */
-	uint32_t reserved[1];
+	/* extended data offset, 0 if no extended data */
```

Comment:
extended data offset **from here**

Reply:
@lgirdwood Indeed, you lost me a bit as well. @keyonjie I think the rationale is to separate the common header from the component specific part -- Liam discusses this in https://github.com/thesofproject/sof/pull/2920/commits/b653465a2149bd88ee6afd834ad781487f231c2f#r446062568 . sof_ipc_comp struct covers the whole message (including the common part), so if ext_data_offset is relative to address of sof_ipc_comp, then the layers are mixed and FW code dealing with this has to have access to the whole IPC struct. But this is a bit hypothetical now when the component specific code in handler.c (like ipc_glb_tplg_comp_new()) always gets the full message. OTOH, "from here" offset works as well -- it's not breaking anything either.

Reply:
@keyonjie yes I'm serious and it's what I meant in our conversation.. This is the common use case of offsets in code, it's always relative to the "offset" variable as it makes an easy addition to get the next data item.

Reply:
>	/* extended data position is relative to the offset member i.e. "from here" */ 
	comp_ext = (struct sof_ipc_comp_new_ext *)
		   ((uint8_t *)(comp->ext_data_offset) + comp->ext_data_offset);

This should be corrected like below?
```
	comp_ext = (struct sof_ipc_comp_new_ext *)
		   ((uint8_t *)(&comp->ext_data_offset) + comp->ext_data_offset);
```
Actually I don't think that is "an easy addition" comparing with my exited one as below :
```
		comp_ext = (struct sof_ipc_comp_new_ext *)
			   ((uint8_t *)comp + comp->ext_data_offset);
```

@lgirdwood I am fine with either definitions, though the new one you suggested is not very straight-forward to be commented ("specific_members_size + 4" as @marc-hb commented? or "extended data offset from here"?), and it's actually quite error-prone as it has an odd 4 bytes addition included. You did suggest the offset based on the whole IPC struct in the very begin of your proposing of the appending extended data solution mails, but if you stick to the new idea here you mentioned above, I can change it, just to speed up the alignment of the long-pending PRs,

The one @marc-hb suggested (naming it xx_length) is not good enough since we are using 0 offset to denote that there is no extended data existed.

Reply:
> The naming @marc-hb suggested (naming it xx_length) is not good enough since we are using 0 offset to denote that there is no extended data existed.

If there's no extended data then you don't need the length of what's before it  to know where it is.

NULL pointers don't really make sense either, yet they are very frequently used as a special value.

Negative values are also common for special values; maybe the reserved field should be changed to a signed `int32_t`? That would leave plenty reserved values left.

BTW the latest and greatest `src/include/ipc/header.h` has been merged (PR #3143),  please rebase and edit the `Reserved` field doxygen in that file.